### PR TITLE
feat: Support delegated targets roles in repo writer

### PR DESCRIPTION
**Pull request by mnm678**
This pr is a takeover of #171, which I don't seem to be able to edit directly.

cc @ethan-lowman-dd @asraa @trishankatdatadog @hosseinsia

### CODE REVIEW

File: data/types.go
Comment (asraa) [inline data/types.go:90]: nit: targets and delegated roles

File: local_store.go
Comment (asraa) [inline local_store.go:429]: would loadPrivateKeys possibly return duped keys?

File: local_store.go
Comment (asraa) [inline local_store.go:524]: i think these lines can move outside of the for loop so you only dedupe once as long as you hold a local keyIDsForRole inside the for loop

**Review by asraa — COMMENTED** [review]
Just started doing the review, just got caught up a little in the key generation not being compatible with delegations -- right now i think it is, except for the hashed bin delegations? (i think it should end up saving the keys in a role file, possibly duplicating them)

**Comment by ethan-lowman-dd** [comment]
I haven't looked at the parts of this I wrote in a while, so I'm unfortunately a bit rusty on what's unfinished, but this is the testing script I was using (haven't converted it to unit tests yet): https://gist.github.com/ethan-lowman-dd/cbc403cd0e68215703f903b7b8777c9c

Running this script on the current branch, regular delegated role signing seems to work. Key management for hash bin delegated signing is not working -- the `leaf` signer is not being fetched properly, so no signature is written.

There's also a potential issue with delegation traversal, but I need to read the spec closer to verify -- in the test, A.txt can not be signed since there is no explicit delegation for its path, but I think the top level targets role should sign in that case.

Additionally, I think it's still unaddressed how hash bin delegated keys would be loaded after the delegation is first written. In the test script, we generate the `leaf` key before we use it, so it's cached in memory by key ID. However, since the hash bins role name `leaf` doesn't appear anywhere in TUF metadata (its metadata is named `bins_XX.json`, not `leaf.json`), when we load the repo a second time, we have no handle to the `leaf` key on disk. One way to fix this would be to require clients to manually load delegated keys before signing with them, but that's perhaps not so friendly. I would lean towards something  like storing keys on disk by their key ID rather than their role name, but that's a pretty big breaking change. Whatever the solution, it should ideally work for remote HSM signers too.

**Comment by ethan-lowman-dd** [comment]
Ah @asraa good point, another way to make the disk-backed key store work with delegated roles would be to just write it to disk under each of `bins_00...bins_ff` for example -- `n` copies of the key for `n` bins. That actually might be the easiest.

**Comment by mnm678** [comment]
> Just started doing the review, just got caught up a little in the key generation not being compatible with delegations -- right now i think it is, except for the hashed bin delegations? (i think it should end up saving the keys in a role file, possibly duplicating them)

Yes, I was also running into this [in tuf-notary](https://github.com/notaryproject/tuf/pull/25). You can use the `keys.GenerateEd25519Key()`, etc, but not the helper functions. That would be a good addition here.

**Comment by trishankatdatadog** [comment]
Going to defer to the rest here. Great job, everyone!

File: local_store.go
Comment (mnm678) [reply local_store.go:429]: On another read, I think this shouldn't be an issue with the current design. It's possible that loadPrivateKeys would have duplicates if someone purposefully added them multiple times in different files, but I don't think we need to worry about that here (although we would for verification)

**Comment by mnm678** [comment]
Thanks for the test script @ethan-lowman-dd. I think I fixed the `AddTarget` bug.

File: repo.go
Comment (mnm678) [inline repo.go:510]: Would there ever be a reason to use a version of this function that specifies which delegation to add the target to? Maybe if a higher level delegation decides to sign something instead of having the delegatee do so?

If that's something that might be useful, I can add a function to do so.

Comment (mnm678) [reply repo.go:510]: I went ahead and added this

**Comment by mnm678** [comment]
Could I get another review @ethan-lowman-dd or @asraa? The hashed-bin key storage seems to be working as long as the keys are each stored (`bins_0-1`, etc). There might be a more elegant way to automate this, but I think we can leave that to a future pr.

**Comment by ethan-lowman-dd** [comment]
I pushed some failing tests -- currently, targets metadata that does not have any target files is not signed. I'm currently debugging.

Also, a note for myself -- all of my printf debugging in `repo.go` needs to be removed still.

**Comment by ethan-lowman-dd** [comment]
I fixed the issues I mentioned above. I think it may be okay if the `SetThreshold` modifier does not support delegation for now -- it can probably be a separate PR. I'll try to find some time today or tomorrow to work on more cleanup and tests.

**Comment by trishankatdatadog** [comment]
@lukpueh do you think you can help review?

**Comment by ethan-lowman-dd** [comment]
Any thoughts on whether delegated targets key revocation needs to go in this PR, or whether we can we leave those as TODOs? I wouldn't consider the code useable in production without APIs to revoke keys, but I also don't want to bloat this PR with mostly-independent changes.

**Comment by mnm678** [comment]
> Any thoughts on whether delegated targets key revocation needs to go in this PR, or whether we can we leave those changes as TODOs? I wouldn't consider the code useable in production without APIs to revoke keys, but I also don't want to bloat this PR with mostly-independent changes.

I think we can update all of the add/revoke key functions in a separate pr to keep this one a bit smaller/easier to review. But I agree that we should get those done before a release that includes this pr.

File: local_store.go
Comment (ethan-lowman-dd) [inline local_store.go:233]: @mnm678 This implementation gives precedence to unstaged metadata in the case of staged & unstaged metadata having the same name. Shouldn't staged metadata get priority, as in the implementation of `GetMeta()` on `memoryStore`?

File: repo.go
Comment (rdimitrov) [inline repo.go:322]: Maybe `!roles.IsTopLevelRole(keyRole)`? If I recall correctly, so far it's possible to generate a key (ergo have a keys file that can be passphrase encrypted) only for top-level roles.

File: repo.go
Comment (rdimitrov) [inline repo.go:361]: I see this is moved to `AddVerificationKeyWithExpiration` and so is it okay to verify the role/expiration **after** we do `SaveSigner()`?

**Comment by rdimitrov** [comment]
Just a general comment, but it would be really great for go-tuf if we can try to keep the PRs simpler and independent as much as possible so they are easy to review. For example, If there's some refactoring related to introducing a new feature, try to separate it in a different PR. Trying to make sense of a lot of changes put in a single PR is usually only feasible for people with very in-depth knowledge of the whole codebase. TUF, itself, is not an easy topic, so avoiding this would help in the direction of welcoming more people to review (and even contribute to) the project which would also limit the chance of introducing code that was not thought through well enough 👍

**Comment by ethan-lowman-dd** [comment]
@rdimitrov Since supporting delegated roles touches almost everywhere roles are used, it's been tricky to keep this change compact. However, I think at this point we're confident enough in the necessary refactors that it makes sense to separate them out as you suggested. I'll spend some time on this today.

**Comment by ethan-lowman-dd** [comment]
@mnm678 Mind if I force push a rebase of this branch that includes only the changes not split into the prereq PRs? The new size is `6 files changed, 745 insertions(+), 137 deletions(-)` which is hopefully a lot easier to review.

File: data/types.go
Comment (ethan-lowman-dd) [reply data/types.go]: This is fixed here: https://github.com/theupdateframework/go-tuf/pull/195/files

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:321]: I don't think there is a security impact here, but in https://github.com/theupdateframework/go-tuf/pull/175/commits/00996cacd27ad5591d35e54ce50d93a36708adaa I reordered these calls so the verification key is added (with the given expiration date) before the signer is added.

File: local_store.go
Comment (ethan-lowman-dd) [reply local_store.go]: Is this comment still relevant? I don't think it applies to the current state of the code, since this line is already outside the for loop.

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go]: Fixed.

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:321]: Actually this commit made lots of tests fail. Working on figuring out why.

Comment (ethan-lowman-dd) [reply repo.go:321]: Addressed in https://github.com/theupdateframework/go-tuf/pull/175/commits/a983813a0fb5f631dee3352704b19898e2d00d7a: "Revert 00996cacd, add back arg validation to AddPrivateKeyWithExpires, and add comment about func call order"

File: repo.go
Comment (asraa) [inline repo.go:870]: Thanks for supported prioritized delegations!

**Comment by ethan-lowman-dd** [comment]
@mnm678 When I merged the diffbase, it looks like this PR was closed. Since you own the PR, I'm not able to set the diffbase. Could you set it to `master` please and reopen the PR?

**Comment by ethan-lowman-dd** [comment]
@mnm678 Ah, I figured it out. I just had to temporarily restore the diffbase branch to reopen, then edit the diffbase.

**Comment by ethan-lowman-dd** [comment]
Something weird happened with the rebase and some changes from prereq PRs were reverted. Going to fix those and then squash the commits on this branch.

**Comment by asraa** [comment]
Hey! Now that the pre-reqs are merged, can we update this one and I'll start a thorough review?

**Comment by ethan-lowman-dd** [comment]
@asraa Sorry for the delay. Are there updates you need other than https://github.com/theupdateframework/go-tuf/issues/239? My plan was to finish the TODOs marked in comments here and add tests where missing.

**Comment by asraa** [comment]
> @asraa Sorry for the delay. Are there updates you need other than https://github.com/theupdateframework/go-tuf/issues/239? My plan was to finish the TODOs marked in comments here and add tests where missing.

I think we can tackle that separately -- imo if people agree that we need easier ways to update fields (versions/expiration) for any data.Signed than this could go in a more general PR.

File: repo.go
Comment (ethan-lowman-dd) [inline repo.go:519]: I think it might be nice to have a more compact set of APIs for manipulating targets, using the options pattern to customize expiration, preferred delegated target role, etc. This is out of scope for this PR, though.

File: repo.go
Comment (ethan-lowman-dd) [inline repo.go:1035]: I'm not sure this implementation is correct. I think in order to produce metadata that's verifiable by any valid incoming delegation key set, the metadata needs to be signed by the union of the set of all keys in all the delegations that lead to this role. Currently, I think this is only signing using keys for the first traversed delegation that leads to the `preferredRole` (or if `preferredRole` is not specified, the first delegation that can sign for the given path).

Comment (ethan-lowman-dd) [reply repo.go:1035]: This is not applicable if the delegation graph is a tree, but the more general case where a role can have multiple valid delegation key sets adds significant complexity like this.

File: repo.go
Comment (ethan-lowman-dd) [inline repo.go:742]: While making `AddTargetsWithDigest` support delegated targets, I removed `writeTargetWithExpires` since it was only used in two places (`AddTargetsWithDigest` and `Init`). Versions numbers are explicitly bumped/set everywhere else in the code, and having the side effect of mutating the passed `data.Targets` is slightly confusing with that convention.

**Comment by ethan-lowman-dd** [comment]
@asraa Sorry for all the delays, it's been hard to find time for this. This should be ready for review now.

File: repo.go
Comment (asraa) [inline repo.go:653]: It seems like this removes all delegations of a target. If top level targets delegates to Foo and Bar, we can't remove just the Foo delegation, right?

Another comment about this: will this delete the Foo.json and Bar.json from the store?

File: repo.go
Comment (hosseinsia) [inline repo.go:536]: nite: Make the method more readable. 
```suggestion
func (r *Repo) AddDelegatedRole(delegator string, delegatee data.DelegatedRole, keys []*data.PublicKey) error {
```

File: repo.go
Comment (hosseinsia) [inline repo.go:544]: nit: 
```suggestion
func (r *Repo) AddDelegatedRoleWithExpires(delegator string, delegatee data.DelegatedRole, keys []*data.PublicKey, expires time.Time) error {
```

File: repo.go
Comment (hosseinsia) [inline repo.go:311]: allowing non-tree makes this very complex. See this discussion: https://github.com/theupdateframework/specification/issues/214#issuecomment-1086228742

File: repo.go
Comment (asraa) [inline repo.go:741]: does this mean that `Sign` will only work for top-level roles? I would expect any delegation is allowed to be signed.

File: local_store.go
Comment (ethan-lowman-dd) [reply local_store.go:225]: I just verified that this is not an issue. Target files live within the `targets` directory, and `isMetaFile` is only called on files in the top-level listing of `repository` and `staged`, not a recursive listing. The directory `repository/targets` or `staged/targets` may be passed to `isMetaFile`, but malicious JSON files within such as `targets/root.json` will not be passed.

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:653]: This API will remove all delegations of a target. Is that problematic? My thinking was that this was a minimal starting point to support revocation, and we can add APIs for updates and singular deletes in a separate PR.

This will not delete any of `foo.json, bar.json, x.foo.json, x.bar.json` from the store.

I added a missing test for this method.

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go]: This is easy to support and was an oversight. I added this in https://github.com/theupdateframework/go-tuf/pull/175/commits/189af3ec7bb2d58c4c0e5f3a8049a5a94ae72a66.

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:1139]: Having a different interpretation of the empty targets slice depending on the role name feels more confusing to me than the original unexpected behavior. It's possible for clients to work around the unexpected behavior by not calling `AddTargets*` if the list of targets is empty, but having two different behaviors may lead to more subtle bugs and more complex functionality for us to support long term.

File: local_store.go
Comment (ethan-lowman-dd) [reply local_store.go:46]: > The consequence of failing to filter would be having signatures that don't verify, correct

Correct. 

> It's not a security issue, just an availability one.

I don't think it's a security issue -- more of a usability issue. Signing with untrusted or revoked keys does not present any security risk in TUF, since keys must be explicitly trusted in the client workflow before the signatures count toward a threshold.

> Is the information in LocalStore sufficient to do this filtering

You need a `verify.DB` or something equivalent to do the filtering, which is outside the scope of what `GetSigners` does.

Comment (ethan-lowman-dd) [reply local_store.go:46]: Also note that the root role needs to sign with expired keys sometimes, so there's no universal way to verify whether a key is valid without a lot of context.

File: local_store.go
Comment (ethan-lowman-dd) [reply local_store.go:45]: I don't think it's possible, or necessarily desired. See this comment thread on the same topic: https://github.com/theupdateframework/go-tuf/pull/175/files#r839629591

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:167]: Tracking this issue here: https://github.com/theupdateframework/go-tuf/issues/266

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:312]: Tracking this here: https://github.com/theupdateframework/go-tuf/issues/266

The input validation on the role name happens in `AddPrivateKeyWithExpires`.

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:849]: I think the name `getSignersInDB` is more clear than `getSigners`, since it distinguishes the functionality from `LocalStore.GetSigners`, which returns an unfiltered list of signers in the local store.  `getSignersInDB` filters the returned list of signers to those present in the given `verify.DB`.

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go]: I changed the method name in https://github.com/theupdateframework/go-tuf/pull/175/commits/0f60bf5964080d24bd5ad7c9847b87aaca6f4d9f, but left the `delegatedRole` name, since to me the parallel names `delegator` and `delegatee` look like they should be the same type, but they are in fact different types.

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go]: See comment above and https://github.com/theupdateframework/go-tuf/pull/175/commits/0f60bf5964080d24bd5ad7c9847b87aaca6f4d9f

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go]: I tried to write a unit test for `Sign` and `AddOrUpdateSignature` using only public methods, but couldn't find a way to create a new unsigned metadata file for a delegated role. I think we would need to make `Repo.setMeta` public to support that, and I'm not sure we want to do that in this same PR. Did you have code that needed to use `Sign` or `AddOrUpdateSignature`? If so, how were you creating the unsigned meta files?

Comment (asraa) [reply repo.go]: Yeah! I was able to create unsigned meta files. 

Shouldn't this function do it? https://github.com/theupdateframework/go-tuf/blob/0f60bf5964080d24bd5ad7c9847b87aaca6f4d9f/repo.go#L579

So Init repo, gen top-level keys, add targets and delegated role -- the delegated role addition doesn't include adding the keys to the store, so you'll end up with it unsigned. See example [here](https://github.com/asraa/test-sigstore-root/blob/765a319bd3778a6661e69fc9b816e535b2bf1a0c/cmd/tuf/app/add-delegation.go#L89-L112) (Note: i am using a slightly older API)

Comment (ethan-lowman-dd) [reply repo.go]: In unit tests, previously what I've been doing is calling `localStore.SaveSigner` with the delegated role key before I call `repo.AddDelegatedRole`.

I added unit tests in https://github.com/theupdateframework/go-tuf/pull/175/commits/2d5f10dfd0ea3e5a6b766cc13739a93ffef6e1c8 that test `Sign` and `AddOrUpdateSignature` with the usage pattern you described.

File: client/errors.go
Comment (ethan-lowman-dd) [reply client/errors.go]: Yes, but I went ahead and just deleted the unused functions in https://github.com/theupdateframework/go-tuf/pull/175/commits/cbb2528225e3f789caa15d55481deba1b664b796

File: repo.go
Comment (asraa) [inline repo.go:592]: Do we expect this file to possibly already be created? On one hand I suppose this could be used to "update" a delegation, but on the other hand, this function will also clear any targets that were in the delegated file.

Side question: Does the version of the delegateeFile bump if the signers change, but the content (targets and further delegations) do not?

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:88]: If `data.NewTargets` or `NewRoot()` set `Version` to 1, that breaks a number of existing tests. Currently, the version is incremented the first time the metadata is staged to the local store. Initializing the version to zero is a bit of a hack that gets the version incremented to 1 when the meta file is staged.

I agree this could be more intuitive, but we'd have to change code unrelated to delegations and it would be a breaking change for `data.NewTargets` and `data.NewRoot`. What do you think about creating an issue for this and handling this in a separate PR?

File: repo.go
Comment (ethan-lowman-dd) [reply repo.go:592]: > Do we expect this file to possibly already be created?

`delegateeFile` may already be created. For example, if you have a diamond delegation graph with edges (A->B), (A->C), (B->D), (C->D), `D.json` will already exist when you create the delegation from C to D.

> this function will also clear any targets that were in the delegated file

This function should not clear any targets in `delegateeFile`, since the call `r.targets(delegatee)` several lines up returns existing metadata if it exists, or creates empty targets metadata if necessary. I added a test in https://github.com/theupdateframework/go-tuf/pull/175/commits/be25f9324374ffb1c599d3f801150cfe6a6aa344 to verify this.

> Does the version of the delegateeFile bump if the signers change, but the content (targets and further delegations) do not?

The version number should not change unless the content within the `signed` object changes. Adding or removing from the `signatures` section of a `data.Signed` meta file should not change the version number. However, `AddDelegatedRoleWithExpires` does currently refresh the expiration of the delegatee file (defined inside `signed`), which will trigger a version bump (also verified in https://github.com/theupdateframework/go-tuf/pull/175/commits/be25f9324374ffb1c599d3f801150cfe6a6aa344).

File: repo.go
Comment (asraa) [reply repo.go:87]: Sg! https://github.com/theupdateframework/go-tuf/issues/272

**Review by asraa — APPROVED** [review]
Let's go! Really want to start getting this code in, and we can tackle the cleanups afterwards. If need be let's make a milestone tracking issue for all the tasks we need to do so we can sort out who can own each.

**Review by hosseinsia — APPROVED** [review]
LGTM

File: repo.go
Comment (abs007) [reply repo.go:87]: @ethan-lowman-dd  I'm attempting to solve this issue and I have set `Version` to 1 for `NewRoot()` and `NewTargets()`,.
So I'm wondering if I should remove the code lines that increment the version during staging since `Version` has a value of 1 by default now. Would appreciate your feedback on this.

Comment (ethan-lowman-dd) [reply repo.go:87]: Hello @abs007, let's correspond in #272 please. This is a merged PR. Could you please share on that issue the problem you're having, ideally including code to reproduce?

**Pull request merged**
